urlb44e9843ba_QuasimodoNZ_SWEN222_Game_Project_tgz-pJ8-Game_World_GameWorldLevelBuilderJ8/src/Client_Server/AGClient.java

```diff
@@ -0,0 +1,310 @@
+					tellGamePanelRepaint();
+				} else if (object.getClass() == ChatMessage.class) {
+					ChatMessage cm = (ChatMessage) object;
+					host.appendChatArea(cm.name, cm.message);
+				} else if (object.getClass() == WinningWorld.class) {
+					WinningWorld w = (WinningWorld) object;
+					world = GameWorld.fromByteArray(w.world);
+					world.setWin(w.won);
+					world.setWinner(w.player);
+					tellGamePanelRepaint();
+				}
+			}
+		});
+		registerPlayer(playerName);
+	}
+
+	/**
+	 * Register a new player with the GameWorld hosted by the server.
+	 *
+	 * @param name Player name of the new player
+	 */
+	private void registerPlayer(String name) {
+		NewPlayer np = new NewPlayer();
+		np.name = name;
+		client.sendTCP(np);
+	}
+
+	/**
+	 * Sends a KeyEvent from this client to the server.
+	 * The KeyEvent is sent as in int and interpreted by the server
+	 * based on the int value.
+	 *
+	 * @param ev the KeyEvent of the user to be sent to the server
+	 * @throws IOException
+	 * @throws InterruptedException
+	 */
+	public void sendAction(String name, KeyEvent ev) throws IOException, InterruptedException {
+		Move m = new Move();
+		m.name = name;
+		m.key = ev.getKeyCode();
+		world.move(name, ev.getKeyCode());
+		tellGamePanelRepaint();
+		client.sendTCP(m);
+	}
+
+	/**
+	 * Sends a MouseEvent from this client to the server.
+	 * The MouseEvent is sent as in int (which button was clicked)
+	 * accompanied by the x and y coordinates of the mouse click.
+	 *
+	 * @param ev the MouseEvent of the user to be sent to the server
+	 * @throws IOException
+	 */
+	public void sendAction(String name, MouseEvent ev) throws IOException {
+		Click c = new Click();
+		c.name = name;
+		c.key = ev.getButton();
+		c.x = ev.getX();
+		c.y = ev.getY();
+		client.sendTCP(c);
+	}
+
+	/**
+	 * Sends the given text to all the other chatAreas via there client
+	 *
+	 * @param text message to be sent
+	 * @throws IOException Invalid Text
+	 */
+	public void sendChat(String name, String text) throws IOException {
+		ChatMessage cm = new ChatMessage();
+		cm.name = name;
+		cm.message = text;
+		client.sendTCP(cm);
+	}
+
+	/**
+	 * Moves this player through an ExternalDoor, changing his location.
+	 * Sends the door to the server so the master GameWorld can register
+	 * the change in game state and send it out to everyone.
+	 *
+	 * @param door ExternalDoor to be moved through
+	 */
+	public void moveThroughDoor(ExternalDoor door) {
+		MoveThroughDoor mtd = new MoveThroughDoor();
+		mtd.name = playerName;
+		mtd.door = door;
+		client.sendTCP(mtd);
+	}
+
+	/**
+	 * Attempts to close the InternalDoor. The player, door and action
+	 * are then forwarded to Server to see whether this player has the
+	 * right key to close the door.
+	 *
+	 * @param door InternalDoor trying to be closed.
+	 */
+	public void closeDoor(InternalDoor door) {
+		InteractDoor id = new InteractDoor();
+		id.name = playerName;
+		id.action = 0; //close
+		id.door = door;
+		client.sendTCP(id);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 159:163, 174:178.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.